### PR TITLE
Add override hooks into Presenters

### PR DESCRIPTION
**Pull request by matks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Add override hooks into Presenters, allowing to enrich the data built by these services
| Type?         | improvement
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/11125
| How to test?  | No need for QA I guess 🤔

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/17203)

### CODE REVIEW

**Comment by jf-viguier** [comment]
Thanks @matks, sounds good

File: src/Adapter/Presenter/Cart/CartPresenter.php [hunk 1]
```diff
@@ -477,7 +477,13 @@ function ($voucher) {
                     'Shop.Theme.Checkout'
                 ) :
                 '',
+        ];
+
+        Hook::exec('overridePresentedCart',
```
Comment (PierreRambaud) [inline src/Adapter/Presenter/Cart/CartPresenter.php]: What about https://devdocs.prestashop.com/1.7/modules/concepts/hooks/#naming-scheme :thinking: 
```suggestion
        Hook::exec('actionPresentCart',
```
wdyt?

Comment (matks) [reply src/Adapter/Presenter/Cart/CartPresenter.php]: I saw a 3rd type of hooks here: https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/Presenter/Cart/CartPresenter.php#L423 and it looked good 🤔but I can change the PR if it's better

Comment (PierreRambaud) [reply src/Adapter/Presenter/Cart/CartPresenter.php]: Or update the doc :thinking:

**Comment by 123monsite-regis** [comment]
hi @matks 

do you think, this pull can also be a good way to solve :
https://github.com/PrestaShop/PrestaShop/issues/11901

?

regards

**Comment by matks** [comment]
> hi @matks
> 
> do you think, this pull can also be a good way to solve :
> #11901
> 
> ?
> 
> regards

This is 2 different topics, but it was a good idea to ping me about https://github.com/PrestaShop/PrestaShop/issues/11901. @jocel1 did not have the time to work on it and, to my knowledge, does not work on PrestaShop anymore. I'll take over 😉

**Comment by matks** [comment]
I'll check Travis first then address Pierre's feedback to push this PR forward

**Comment by matks** [comment]
Interesting 🤔 sanity tests fail for php7.3 only
https://travis-ci.com/PrestaShop/PrestaShop/jobs/276241936

Though my code does look php7.3-compliant 😅

**Comment by matks** [comment]
@PierreRambaud Can you please checkout this branch and tell me if something seems wrong on FO (should be easy to spot according to Travis). I need to upgrade my env php to check with php7.3

**Comment by PierreRambaud** [comment]
@matks I can't reproduce the error, could you rebase your branch please?

**Comment by matks** [comment]
PR rebased and I update the hook naming to comply with doc

**Review by PierreRambaud — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by jf-viguier** [comment]
I can't wait for this ! 1.7.7 ?

**Comment by matks** [comment]
Rebased, PrettyCI should be happy

@jf-viguier Pablo (@eternoendless) mapped the issue to 1.7.8 kanban but if he agrees, we could backport this into 1.7.7

**Review by PierreRambaud — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Comment by matks** [comment]
Thanks for review !

**Comment by matks** [comment]
I need to know whether I should
- update hook.xml file
- add new hooks into database in 1.7.7.0 SQL script
@Quetzacoalt91 @eternoendless can you tell me ?

**Comment by jf-viguier** [comment]
Please go to 1.7.7, it's just new hooks but they can save lives ;-)

**Comment by matks** [comment]
> I need to know whether I should
> 
> * update hook.xml file
> * add new hooks into database in 1.7.7.0 SQL script
>   @Quetzacoalt91 @eternoendless can you tell me ?

Got answer: it's yes
See https://github.com/PrestaShop/PrestaShop/pull/17477

**Comment by matks** [comment]
I did an analysis of performance impact.
Only hook that seems to be overused is `actionPresentCart`. It seems to be called multiple times on pages like Checkout, Home, Categories and Product. Number of products in cart have an impact too.

With an empty cart,
- called 10 times on Homepage
- called multiple times on different Category pages, up to 14 times (13 on category "Accessories", 14 on "All products")
- called 3 times on one Product page
- called 4 times on Cart page

On category page it seems to be called for each category shown and each product displayed 🤔 

Cart content seems to modify these numbers a little bit but it's not "one more product = one more call".

**Comment by matks** [comment]
Looks like it's a lot of ajax calls that fire `ModuleFrontControllerCore::initContent()` that fire `FrontControllerCore:: assignGeneralPurposeVariables()` that uses CartPresenter to build the cart

**Comment by jf-viguier** [comment]
@matks thanks for the analysis. The cart calculation system definitely needs a complete rewrite !

**Comment by matks** [comment]
It seems in the end that CartPresenter is not overused in a single HTTP request, however multiple modules, hooked on the same page, will build the Cart although they dont use it, because it is being built using `CartPresenter::present($cart)` .

Looks like it's a lot of ajax calls that fire ModuleFrontControllerCore::initContent() that fire FrontControllerCore:: assignGeneralPurposeVariables()` which is called by main `initContent()`.

The consequences of this are:
some front-office routes and module routes build the Cart although they dont use it => waste of CPU, memory, response time and possibly useless SQL queries being performed

**Pull request merged**